third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/op_equal/valid.move

```diff
@@ -0,0 +1,166 @@
+//# publish
+module 0x42::test {
```

Comment:
It turns out that in Rust, the RHS of the `+=` operator is evaluated first.
> If both types are primitives, then the modifying operand will be evaluated first followed by the assigned operand. It will then set the value of the assigned operand’s place to the value of performing the operation of the operator with the values of the assigned operand and modifying operand.
https://doc.rust-lang.org/reference/expressions/operator-expr.html#compound-assignment-expressions

We are aligned with Rust now. The Rust version of the test is here for convenience
```
fn test0() -> u64 {
    let mut v = 1;
    v += {
        v += {
            v += 2;
            v
        };
        v
    };
    v
}

fn test1() -> u64 {
    let mut v = 1;
    v += {
        v += 2;
        v
    };
    v
}

fn mod1(r: &mut u64) {
    *r += 2;
}

fn test2() -> u64 {
    let mut v = 1;
    v += {
        mod1(&mut v);
        v
    };
    v
}

fn mod2(r: &mut u64) -> u64 {
    *r += 2;
    *r
}

fn test3() -> u64 {
    let mut v = 1;
    v += mod2(&mut v);
    v
}

fn main() {
    println!("{:?}", test0());
    println!("{:?}", test1());
    println!("{:?}", test2());
    println!("{:?}", test3());
}
```

Reply:
Nice.

I saw that you added the tests I posted under bullet point 1, but don't see the tests I suggested in bullet points 2 and 3. If you have already added them, can you point them to me? If not, could you please add them (and point me to them)?

Reply:
Bullet point2: `third_party/move/move-compiler-v2/tests/op-equal/valid2.exp`
It however compiles in the latest commit, because of the change in semantics. The corresponding Rust program also compiles
```rust
struct S {
    x: u64,
}

impl S {
    fn foo(self: &mut S) -> u64 {
        self.x += 1;
        1
    }
}

fn test() -> u64 {
    let mut s = S { x: 0 };
    let p = &mut s.x;
    *p = *p + 
    s.x += s.foo();
    s.x
}
```
Bullet point3: `third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/op_equal/no_double_eval.move`